### ISSUE TITLE
Digraphs not compiling in Centos 7

### ISSUE THREAD

**Issue by enriqueartal** [gap-packages/Digraphs#297]
In Centos 7, gcc 485, Digraphs fail to compile with this error:

~~~~
gmake[1]: Entering directory `/usr/local/gap/pkg/digraphs-1.1.1'
  CC       src/digraphs_la-digraphs.lo
src/digraphs.c: In function ‘FuncOutNeighboursFromSourceRange’:
src/digraphs.c:102:3: error: ‘for’ loop initial declarations are only allowed in C99 mode
for (Int i = 1; i <= n; i++) {
^
src/digraphs.c:102:3: note: use option -std=c99 or -std=gnu99 to compile your code
src/digraphs.c:109:12: error: redefinition of ‘i’
for (Int i = 1; i <= LEN_LIST(src); i++) {
^
src/digraphs.c:102:12: note: previous definition of ‘i’ was here
for (Int i = 1; i <= n; i++) {
^
src/digraphs.c:109:3: error: ‘for’ loop initial declarations are only allowed in C99 mode
for (Int i = 1; i <= LEN_LIST(src); i++) {
^
src/digraphs.c: In function ‘DigraphNrEdges’:
src/digraphs.c:126:5: error: ‘for’ loop initial declarations are only allowed in C99 mode
for (Int i = 1; i <= n; i++) {
^
src/digraphs.c: In function ‘FuncDIGRAPH_TRANS_REDUCTION’:
src/digraphs.c:502:3: error: ‘for’ loop initial declarations are only allowed in C99 mode
for (UInt i = 1; i <= n; i++) {
^
src/digraphs.c:515:13: error: redefinition of ‘i’
for (UInt i = 1; i <= n; i++) {
^
src/digraphs.c:502:13: note: previous definition of ‘i’ was here
for (UInt i = 1; i <= n; i++) {
^
src/digraphs.c:515:3: error: ‘for’ loop initial declarations are only allowed in C99 mode
for (UInt i = 1; i <= n; i++) {
^
src/digraphs.c:573:11: error: ‘for’ loop initial declarations are only allowed in C99 mode
for (UInt m = 0; m < n; m++) {
^
src/digraphs.c: In function ‘buildBlissDigraph’:
src/digraphs.c:1613:7: error: ‘for’ loop initial declarations are only allowed in C99 mode
for (Int l = 1; l <= len; l++) {
^
src/digraphs.c: In function ‘MultiplyList’:
src/digraphs.c:1756:3: error: ‘for’ loop initial declarations are only allowed in C99 mode
for (int i = 0; i < length; ++i) {
^
gmake[1]: Leaving directory `/usr/local/gap/pkg/digraphs-1.1.1'
gmake[1]: *** [src/digraphs_la-digraphs.lo] Error 1
gmake: *** [all-recursive] Error 1

WARNING: Failed to build digraphs-1.1.1
~~~~

**Comment by james-d-mitchell** [comment]
Thanks for the report @enriqueartal, could you possibly check if adding the c flag `-std=c99`, resolves this? 

~~~~
./configure CFLAGS="-std=c99"
~~~~

**Comment by enriqueartal** [comment]
Thanks for the quick answer. With this option the error is:
  CC       src/digraphs_la-digraphs.lo
In file included from /usr/local/gap/src/gapstate.h:18:0,
                 from /usr/local/gap/src/code.h:19,
                 from /usr/local/gap/src/gap_all.h:25,
                 from /usr/local/gap/src/compiled.h:26,
                 from src/digraphs.h:19,
                 from src/digraphs.c:15:
/usr/local/gap/src/sysjmp.h:30:19: error: unknown type name 'sigjmp_buf'
 #define syJmp_buf sigjmp_buf
                   ^
/usr/local/gap/src/sysjmp.h:41:16: note: in expansion of macro 'syJmp_buf'
 void syLongjmp(syJmp_buf * buf, int val) NORETURN;
                ^
In file included from /usr/local/gap/src/code.h:19:0,
                 from /usr/local/gap/src/gap_all.h:25,
                 from /usr/local/gap/src/compiled.h:26,
                 from src/digraphs.h:19,
                 from src/digraphs.c:15:
/usr/local/gap/src/gapstate.h:53:5: error: unknown type name 'sigjmp_buf'
     syJmp_buf ReadJmpError;
     ^
make[1]: *** [src/digraphs_la-digraphs.lo] Error 1
make[1]: Leaving directory `/usr/local/gap/pkg/digraphs-1.1.1'
make: *** [all-recursive] Error 1

**Comment by james-d-mitchell** [comment]
No problem, thanks for the report. Do you have the possibility to use a newer version of GCC? Version 4.8.5 is nearly 5 years old, and I suspect that with a newer version, this issue will disappear.

Also note that the new error you give above, seems to indicate that GAP itself won't compile, is that right?

**Comment by enriqueartal** [comment]
Gap compiled and most of the packages (should I recompile gap with the CFLAGS option?). Everything goes OK in a Fedora 31 system with gcc9. I do not know if it is safe to change the gcc version in the Centos 7 server, I have only network access to it.

**Comment by james-d-mitchell** [comment]
Which version of GAP are you using? GAP 4.10 for example doesn't have the file `src/sysjmp.h`, which appears to be where the error is coming from. Just to double check, GAP itself compiled ok using gcc 4.8.5 but Digraphs did not?

**Comment by james-d-mitchell** [comment]
It shouldn't be necessary to recompile GAP using the CFLAGS option.

**Comment by enriqueartal** [comment]
GAP itself compiled ok using gcc 4.8.5 but Digraphs did not; the version is the development version. My original plan was to install gap_packages in Sage, where I got the original error. Doing `CFLAGS="-std=c99" ./sage -i gap_packages` I bypass the first problem but not the second.

**Comment by james-d-mitchell** [comment]
It might be the case that there are some changes in the GAP development version that break Digraphs, and that's what you are discovering here. Can you use one of the released versions of GAP instead?

**Comment by enriqueartal** [comment]
No problem for GAP but the problem exists when working in sage which does not use the dev version.

**Comment by james-d-mitchell** [comment]
I'm not sure I understand your final comment. I don't really know how the sage setup works with GAP, the issue can't exist with GAP 4.10 since the file reporting the error doesn't exist in GAP 4.10. I can add something to the autotools set up to check that C99 is enabled. To do anything further, I think we'd need more details of exactly what is and isn't working.

**Comment by enriqueartal** [comment]
Sorry, what I mean is if I try to install gap_packages in Sage (which includes digraphs-1.1.1, gap version is 4.10.2) I get the the following error:

make[3]: Entering directory `/usr/local/sage/local/var/tmp/sage/build/gap_packages-4.10.2.p1/src/pkg/digraphs-0.15.2'
/usr/bin/bash ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I./src  -I/usr/local/sage/local/share/gap/gen -I/usr/local/sage/local/share/gap/src -I/usr/local/sage/local/share/gap -DHAVE_CONFIG_H -I/usr/local/sage/local/include         -O3 -march=native -mpopcnt -O2 -g   -I./extern/edge-addition-planarity-suite-Version_3.0.0.5/ -I./extern/  -std=c99 -c -o src/digraphs_la-digraphs.lo `test -f 'src/digraphs.c' || echo './'`src/digraphs.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I./src -I/usr/local/sage/local/share/gap/gen -I/usr/local/sage/local/share/gap/src -I/usr/local/sage/local/share/gap -DHAVE_CONFIG_H -I/usr/local/sage/local/include -O3 -march=native -mpopcnt -O2 -g -I./extern/edge-addition-planarity-suite-Version_3.0.0.5/ -I./extern/ -std=c99 -c src/digraphs.c  -fPIC -DPIC -o src/.libs/digraphs_la-digraphs.o
In file included from /usr/local/sage/local/share/gap/src/gasman.h:39:0,
                 from /usr/local/sage/local/share/gap/src/objects.h:20,
                 from /usr/local/sage/local/share/gap/src/ariths.h:17,
                 from /usr/local/sage/local/share/gap/src/compiled.h:17,
                 from src/digraphs.h:19,
                 from src/digraphs.c:15:
/usr/local/sage/local/share/gap/src/system.h:511:19: error: unknown type name 'sigjmp_buf'
 #define syJmp_buf sigjmp_buf
                   ^
/usr/local/sage/local/share/gap/src/system.h:522:16: note: in expansion of macro 'syJmp_buf'
 void syLongjmp(syJmp_buf* buf, int val) NORETURN;
                ^
In file included from /usr/local/sage/local/share/gap/src/compiled.h:33:0,
                 from src/digraphs.h:19,
                 from src/digraphs.c:15:
/usr/local/sage/local/share/gap/src/gapstate.h:51:5: error: unknown type name 'sigjmp_buf'
     syJmp_buf ReadJmpError;
     ^
make[3]: *** [src/digraphs_la-digraphs.lo] Error 1
make[3]: Leaving directory `/usr/local/sage/local/var/tmp/sage/build/gap_packages-4.10.2.p1/src/pkg/digraphs-0.15.2'
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory `/usr/local/sage/local/var/tmp/sage/build/gap_packages-4.10.2.p1/src/pkg/digraphs-0.15.2'
********************************************************************************
Error building gap_packages-4.10.2.p1

**Comment by james-d-mitchell** [comment]
Thanks for the further info. @fingolfin @ChrisJefferson any ideas what might be causing this?

**Comment by ChrisJefferson** [comment]
Question, do you have the same problem if you build GAP entirely seperately from Sage (get the standard GAP download from www.gap-system.org, not the one packaged inside Sage)?

**Comment by enriqueartal** [comment]
I got the same error with the dev version (no problems with GAP and with most of the packages). With digraphs, either the c99 error or the other one. I will let you know what happens with the last release.

**Comment by ChrisJefferson** [comment]
I'm quickly installing centos 7, to see if I can reproduce it.

**Comment by james-d-mitchell** [comment]
Thanks @ChrisJefferson !

**Comment by ChrisJefferson** [comment]
You need to pass `-std=gnu99` instead of `-std=c99`.

**Comment by ChrisJefferson** [comment]
I think the problem here is a general GAP building problem, rather than a digraphs problem -- GAP uses this flag (`-std=gnu99`) when building on centos, but it doesn't seem to get passed to packages automatically. Other packages don't need it (as they aren't using c99 features).

The `setjmp` related mistakes are just a confusion -- passing `-std=c99` means (to GCC) both "use c99 features" (what we want), and "disable some gnu extensions" (which we don't want), which breaks setjmp. `-std=gnu99` means "enable c99, and keep the gnu extensions you were already using".

**Comment by james-d-mitchell** [comment]
Thanks @ChrisJefferson, does this mean that using the `-std=gnu99` explicitly would fix this issue for now?

**Comment by ChrisJefferson** [comment]
I wouldn't add it into Digraphs yet, as there might be some other computer/system it upsets (GAP only adds it when it needs it). Lets see if @fingolfin has an opinion on if this should be handled by the build script (I think it should)

**Comment by alex-konovalov** [comment]
Is this something that might be already fixed in GAP 4.11?

**Comment by ChrisJefferson** [comment]
My mistake, I think it is fixed in 4.11.

**Comment by fingolfin** [comment]
There are a lot of comments here already, so I may have missed something, but...

If Digraphs needs C99 features, then it is up to the Digraphs build system to add the required compiler flags. Note that GAP does exactly that for its own builds. It is possible that in some circumstances we accidentally leak this compiler setting through to packages (e.g. via GAP_CFLAGS in `sysinfo.gap`) but that's not really intentional: packages that need C99 or C11 should really ensure they set the relevant flags, and not rely on GAP. If using autoconf, this can be done by using the `AC_PROG_CC_C99` macro ([see here for details and gotchas](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/autoconf.html#index-AC_005fPROG_005fCC_005fC99-886)

**Comment by ChrisJefferson** [comment]
Sorry @fingolfin . Everything is (I think) fine in new releases of GAP -- the c99 flags are inherited from GAP correctly (which now also requires C99).

**Comment by enriqueartal** [comment]
Thanks for your help. The compilation works for gap; for sage, still waiting since it compiles gcc first.

**Comment by enriqueartal** [comment]
The package gap_packages of sage has been installed successfully; I was wrong, digraphs version was 0.15.2
I have really appreciated your help. Thanks again, Enrique.

**Comment by enriqueartal** [comment]
I did not close the issue since I do not know if you are planning to make slight changes, but for me it is OK

**Comment by james-d-mitchell** [comment]
No problem, I intend to add a check for C99 to the build setup, and then I'll close the issue, thanks for the report, and for everyone's comments and help.

**Comment by james-d-mitchell** [comment]
Resolved in v1.1.2
